klippa/src/serialize.rs

```diff
@@ -0,0 +1,394 @@
+    }
+}
+
+impl PartialEq for SerializeErrorFlags {
+    fn eq(&self, other: &Self) -> bool {
+        self.0 == other.0
+    }
+}
+
+impl std::ops::BitOr for SerializeErrorFlags {
+    type Output = Self;
+
+    /// Returns the union of the two sets of flags.
+    #[inline]
+    fn bitor(self, other: SerializeErrorFlags) -> Self {
+        Self(self.0 | other.0)
+    }
+}
+
+impl std::ops::BitOrAssign for SerializeErrorFlags {
+    /// Adds the set of flags.
+    #[inline]
+    fn bitor_assign(&mut self, other: Self) {
+        self.0 |= other.0;
+    }
+}
+
+impl std::ops::Not for SerializeErrorFlags {
+    type Output = bool;
+    #[inline]
+    fn not(self) -> bool {
+        self == SerializeErrorFlags::SERIALIZE_ERROR_NONE
+    }
+}
+
+#[allow(dead_code)]
+#[derive(Default)]
+enum OffsetWhence {
+    #[default]
+    Head,
+    Tail,
+    Absolute,
+}
+
+// the maximal len of vec in the object_pool is not expected to exceed u32::MAX
+const NULL_POOL_IDX: usize = u32::MAX as usize;
+
+type PoolIdx = usize;
+#[allow(dead_code)]
+struct Object {
+    head: usize,
+    tail: usize,
+    real_links: Vec<Link>,
+    virtual_links: Vec<Link>,
+    next_obj: PoolIdx,
+}
+
+impl Default for Object {
+    fn default() -> Object {
+        Object {
+            head: 0,
+            tail: 0,
+            real_links: vec![],
+            virtual_links: vec![],
+            next_obj: NULL_POOL_IDX,
+        }
+    }
+}
+
+#[allow(dead_code)]
+#[derive(Default)]
+enum LinkWidth {
+    #[default]
+    Two,
+    Three,
+    Four,
+}
+
+type ObjIdx = u32;
+#[allow(dead_code)]
+const INVALID_OBJ_IDX: ObjIdx = ObjIdx::MAX;
+
+#[allow(dead_code)]
+#[derive(Default)]
+struct Link {
+    width: LinkWidth,
+    is_signed: bool,
+    whence: OffsetWhence,
+    bias: u32,
+    position: u32,
+    objidx: ObjIdx,
+}
+
+#[derive(Default)]
+#[allow(dead_code)]
+pub(crate) struct Serializer {
+    pub(crate) start: usize,
+    pub(crate) end: usize,
+    pub(crate) head: usize,
+    pub(crate) tail: usize,
+    // TODO: zerocopy, debug_depth
+    pub(crate) errors: SerializeErrorFlags,
+
+    data: Vec<u8>,
+    object_pool: ObjectPool,
+    // index for current Object in the object_pool
+    current: PoolIdx,
+
+    packed: Vec<ObjIdx>,
+    packed_map: FnvHashMap<usize, ObjIdx>,
+}
+
+#[allow(dead_code)]
+impl Serializer {
+    pub(crate) fn new(size: u32) -> Self {
+        let buf_size = size as usize;
+        let mut this = Serializer {
+            data: vec![0; buf_size],
+            end: buf_size,
+            tail: buf_size,
+            current: NULL_POOL_IDX,
+            packed: Vec::new(),
+            ..Default::default()
+        };
+
+        this.packed.push(INVALID_OBJ_IDX);
+        this
+    }
+
+    // Embed a single Scalar type
+    pub(crate) fn embed(&mut self, obj: impl Scalar) -> Result<usize, SerializeErrorFlags> {
+        let raw = obj.to_raw();
+        let bytes = raw.as_ref();
+        let size = bytes.len();
```

Comment:
If you have acess to the type of `obj` (by using a named generic parameter, instead of `impl Trait`) than you can get this length with `T::RAW_BYTE_LEN`.